js/ui/codingManager.js

```diff
@@ -0,0 +1,1085 @@
+    let {
+        button,
+        rect,
+        startAngle,
+        finishAngle,
+        startOpacity,
+        finishOpacity,
+        opacityDelay,
+        onRotationComplete,
+        onButtonFadeComplete
+    } = props;
+
+    // this API is deprecated but the best option here to
+    // animate around a point outside of the actor
+    button.rotation_center_y = new Clutter.Vertex({
+        x: button.width - (rect.width * 0.5),
+        y: button.height * 0.5,
+        z: 0
+    });
+    button.rotation_angle_y = startAngle;
+    button.opacity = startOpacity;
+    Tweener.addTween(button, {
+        rotation_angle_y: finishAngle,
+        time: WINDOW_ANIMATION_TIME * 4,
+        transition: 'easeOutQuad',
+        onComplete: function() {
+            if (onRotationComplete) {
+                onRotationComplete();
+            }
+        }
+    });
+    Tweener.addTween(button, {
+        opacity: finishOpacity,
+        time: WINDOW_ANIMATION_TIME * 2,
+        transition: 'linear',
+        delay: opacityDelay,
+        onComplete: function() {
+            if (onButtonFadeComplete) {
+                onButtonFadeComplete();
+            }
+        }
+    });
+}
+
+const WindowTrackingButton = new Lang.Class({
+    Name: 'WindowTrackingButton',
+    Extends: GObject.Object,
+    Properties: {
+        window: GObject.ParamSpec.object('window',
+                                         '',
+                                         '',
+                                         GObject.ParamFlags.READWRITE |
+                                         GObject.ParamFlags.CONSTRUCT_ONLY,
+                                         Meta.Window),
+        builder_window: GObject.ParamSpec.object('builder-window',
+                                                 '',
+                                                 '',
+                                                 GObject.ParamFlags.READWRITE,
+                                                 Meta.Window)
+    },
+    Signals: {
+        'clicked': {}
+    },
+
+    _init: function(params) {
+        this.parent(params);
+
+        // The button will be auto-added to the manager. Note that in order to
+        // remove this button, you will need to call destroy() from outside
+        // this class or use removeChrome from within it.
+        this._button = _createInputEnabledViewSourceButtonInRectCorner(this.window.get_frame_rect());
+        this._button.connect('clicked', Lang.bind(this, function() {
+            this.emit('clicked');
+        }));
+
+        // Connect to signals on the window to determine when to move
+        // hide, and show the button. Note that WindowTrackingButton is
+        // constructed with the primary app window and we listen for signals
+        // on that. This is because of the assumption that both the app
+        // window and builder window are completely synchronized.
+        this._positionChangedId = this.window.connect(
```

Comment:
This coding style is a bit odd; usually if we need to break a line like this, we do either
```
this._signalId = object.connect('signal-name',
    Lang.bind(this, this._callback));
```

or 

```
this._signalId =
    object.connect('signal-name', Lang.bind(this, this._callback));
```